libsolutil/LP.cpp

```diff
@@ -0,0 +1,593 @@
+
+	_tableau.data[0] = originalObjective;
+	for (size_t i = 1; i < rows; ++i)
+		_tableau.data[i].resize(columns);
+
+	//cout << "Tableau after Phase I: " << endl;
+	//printTableau(_tableau);
+
+	// TODO do we need to select a basis?
+	//_tableau = selectLastVectorsAsBasis(move(_tableau));
+	//cout << "After basis correction: " << endl;
+	//printTableau(_tableau);
+
+	return make_pair(LPResult::Feasible, move(_tableau));
+}
+
+bool needsPhaseI(Tableau const& _tableau)
+{
+	for (size_t i = 1; i < _tableau.data.size(); ++i)
+		if (_tableau.data[i][0] < 0)
+			return true;
+	return false;
+}
+
+/// Solve the LP Ax <= b s.t. min c^Tx
+/// The first row encodes the objective function
+/// The first column encodes b
+pair<LPResult, vector<rational>> simplex(Tableau _tableau)
+{
+	//printTableau(_tableau);
+	_tableau = toEquationalForm(move(_tableau));
+	//cout << "Equational: " << endl;
+	//printTableau(_tableau);
+	if (needsPhaseI(_tableau))
+	{
+		LPResult result;
+		tie(result, _tableau) = simplexPhaseI(move(_tableau));
+		if (result == LPResult::Infeasible || result == LPResult::Unknown)
+			return make_pair(result, vector<rational>{});
+		solAssert(result == LPResult::Feasible, "");
+	}
+	LPResult result;
+	vector<rational> optimum;
+	tie(result, _tableau) = simplexEq(move(_tableau));
+	if (result == LPResult::Feasible || result == LPResult::Unbounded)
+	{
+		optimum = optimalVector(_tableau);
+		//cout << "Solution: " << endl;
+		//printVector(optimum);
+	}
+	return make_pair(result, move(optimum));
+}
+
+}
+
+void LPSolver::reset()
+{
+	m_state = stack<State>{{State{}}};
+}
+
+void LPSolver::push()
+{
+	m_state.push(m_state.top());
+}
+
+void LPSolver::pop()
+{
+	m_state.pop();
+	solAssert(!m_state.empty(), "");
+}
+
+void LPSolver::declareVariable(string const& _name, SortPointer const& _sort)
+{
+	// TODO This will not be an integer variable in our model.
+	// Introduce a new kind?
+	solAssert(_sort && _sort->kind == Kind::Int, "");
+	solAssert(!m_state.top().variables.count(_name), "");
+	size_t index = m_state.top().variables.size() + 1;
+	m_state.top().variables[_name] = index;
+}
+
+void LPSolver::addAssertion(Expression const& _expr)
+{
+	if (_expr.name == "and")
+	{
+		addAssertion(_expr.arguments.at(0));
+		addAssertion(_expr.arguments.at(1));
+	}
+	else if (_expr.name == "<=")
+	{
+		optional<vector<rational>> left = parseLinearSum(_expr.arguments.at(0));
+		optional<vector<rational>> right = parseLinearSum(_expr.arguments.at(1));
+		if (left && right)
+		{
+			vector<rational> constraint = *left - *right;
+			constraint[0] *= -1;
+			m_state.top().constraints.emplace_back(move(constraint));
+		}
+		else
+		{
+			//cout << "LPSolver: Adedd non-linear constraint." << endl;
+			m_state.top().encounteredUnknownSituation = true;
+		}
+	}
+	else if (_expr.name == ">=")
+		addAssertion(_expr.arguments.at(1) <= _expr.arguments.at(0));
+	else if (_expr.name == "<")
+		addAssertion(_expr.arguments.at(0) <= _expr.arguments.at(1) - 1);
```

Comment:
We should check for underflow here.

Reply:
Why should we? if we have `x < 0` then it results in `x <= -1` which is just infeasible.

Reply:
The problem is that if we encode the following AST

```
let y := calldataload(0)
let x := lt(x, 0)
```

This would end up with:

```
0 <= y <= 2**256 - 1
0 <= x <= -1
```
Which would be incorrect for our uses, because anything you add to this constraint from now will always be infeasible.

So generally, encoding Yul ast should always lead to a feasible system. We should only get infeasibility by adding additional constraints that are not from the Yul AST.

Reply:
I'm probably wrong about the above comment. Let me check how we end up using `<`.

Reply:
When we visit `let x := lt(y, 0)`, we do not add `y < 0` directly, we first query for infeasibility. We only add it inside `if x { ... }`.